File: libserver/smart_contract/connectEthereum.js
let Web3 = require("web3");
let Tx = require("ethereumjs-tx");

//合约地址
let contractAddr = "0xa446b9aBBC394C5c41D4C46FACf18CDd8A64bc11";
//创建web3对象
let web3 = new Web3;

//连接到ropsten
let INFURA_api = "https://goerli.infura.io/v3/d9651a5bfd4548ba8c331e7723b79232";
web3.setProvider(new Web3.providers.HttpProvider(INFURA_api));

//账户地址
let fromAddr = "0x57128a8c478B3fEab65866a9c39d06408c243ce9";

let count = web3.eth.getTransactionCount(fromAddr);
let gasPrice = web3.eth.gasPrice;
let gasLimit = 90000;

let rawTransaction = {
    "from": fromAddr,
    "nonce": web3.toHex(count),
    "gasPrice": web3.toHex(gasPrice),
    "gasLimit": web3.toHex(gasLimit),
    "to": contractAddr, 
    "value": "0x10000", 
    "data": "0x460968dd" //函数名callme取keccak后的bytes4
};

//账户私钥，不包含'0x'字符
let priv_key = new Buffer.from("0a09c8a1913273f24f7c7d610d1a11b0440885300801226c4d39222a32419b9f", "hex");
let tx = new Tx(rawTransaction);

//用私钥签名交易信息
tx.sign(priv_key);
let serializedTx = tx.serialize();

//发送交易
web3.eth.sendRawTransaction('0x'+serializedTx.toString('hex'),
    function(err, hash) {
        if (!err)
            console.log(hash);
        else
            console.log(err);
    });

File: libserver/smart_contract/connectEthereum.py
# -*- coding:utf-8 -*-

from web3 import Web3, HTTPProvider

true = True
false = False
contract_addr="0x090cf53dEe2C6ab1B8d5ad02B08e60fcdae728A1" # 合约地址
config = {
    "abi":[
	{
		"inputs": [],
		"name": "transderToContract",
		"outputs": [
			{
				"internalType": "bool",
				"name": "",
				"type": "bool"
			}
		],
		"stateMutability": "payable",
		"type": "function"
	},
	{
		"inputs": [
			{
				"internalType": "bool",
				"name": "prover",
				"type": "bool"
			},
			{
				"internalType": "address",
				"name": "_to",
				"type": "address"
			}
		],
		"name": "transfer",
		"outputs": [
			{
				"internalType": "bool",
				"name": "",
				"type": "bool"
			}
		],
		"stateMutability": "payable",
		"type": "function"
	},
	{
		"stateMutability": "payable",
		"type": "receive"
	},
	{
		"stateMutability": "payable",
		"type": "fallback"
	},
	{
		"inputs": [
			{
				"internalType": "address",
				"name": "addr",
				"type": "address"
			}
		],
		"name": "getBalance",
		"outputs": [
			{
				"internalType": "uint256",
				"name": "",
				"type": "uint256"
			}
		],
		"stateMutability": "view",
		"type": "function"
	}
],
    "address":contract_addr 
}

INFURA_api = "https://goerli.infura.io/v3/d9651a5bfd4548ba8c331e7723b79232" # INFURA的goerli API地址

web3 = Web3(HTTPProvider(INFURA_api))
contract_instance = web3.eth.contract(address=config['address'], abi=config['abi'])


my_addr = "0xAa3d7608ed56FCbE7aA5c75a5af180eB9Fd0133e" # 协议部署方地址

priv_key = "0xfabde6ac6b55fc46677d9453545e699b0cfe3a244a4418e357c97e6a3f56ec9d" # 协议部署方私钥

#print(web3.isConnected())

def SendTxn(txn,priv_key):
    signed_txn = web3.eth.account.signTransaction(txn,private_key=priv_key)
    try:
    	res  = web3.eth.sendRawTransaction(signed_txn.rawTransaction).hex()
    except ValueError:
    	print("insufficient funds for gas * price + value")
    	return "error"
    txn_receipt = web3.eth.waitForTransactionReceipt(res)
    return "success"
 
def SendTxnTransfer(other_addr):
	before_other_addr_balance = web3.eth.getBalance(other_addr)
	before_contract_balance = web3.eth.getBalance(contract_addr)
	txn = contract_instance.functions.transfer(true,other_addr).buildTransaction(
	    {
		'chainId':5,
		'nonce':web3.eth.getTransactionCount(my_addr),
		'gas':76000,
		'value':Web3.toWei(0,'wei'),
		'gasPrice':web3.eth.gasPrice,
	    }
	)
	print(SendTxn(txn,priv_key))
	print("address : "+other_addr+" before : "+str(before_other_addr_balance)+" after : "+str(web3.eth.getBalance(other_addr)))
	print("address : "+contract_addr+" before : "+str(before_contract_balance)+" after : "+str(web3.eth.getBalance(contract_addr)))
	
def SendTxnTransferToContract(my_addr,value,priv_key):
	before_my_addr_balance = web3.eth.getBalance(my_addr)
	before_contract_balance = web3.eth.getBalance(contract_addr)
	txn = contract_instance.functions.transderToContract().buildTransaction(
	    {
	    	'from':my_addr,
		'chainId':5,
		'nonce':web3.eth.getTransactionCount(my_addr),
		'gas':76000,
		'value':Web3.toWei(int(value),'wei'),
		'gasPrice':web3.eth.gasPrice,
	    }
	)
	print(SendTxn(txn,priv_key))
	print("address : "+my_addr+" before : "+str(before_my_addr_balance)+" after : "+str(web3.eth.getBalance(my_addr)))
	print("address : "+contract_addr+" before : "+str(before_contract_balance)+" after : "+str(web3.eth.getBalance(contract_addr)))
	
def SendTxnGetBalance(other_addr):
	print(web3.eth.getBalance(other_addr))
	
